### ISSUE TITLE
[Bug] No module named dartie (snakemake)

### ISSUE THREAD

**Issue by marsdenl** [gustaveroussy/sopa#121]
Hey there,

I've been trying to use the snakemake workflow for sopa with baysor segmentation. I followed the installation steps specific to snakemake and verified the correct installation of sopa and snakemake (sopa --help and snakemake --help) and everything looked good. However when I tried to run it using the following code:
snakemake --config data_path=PATH\to\region_0 --configfile=PATH\to\baysor_vizgen.yaml --cores 1 --use-conda

I get ModuleNotFoundError: No module named 'datrie'

I realise there have been issues with snakemake and datrie. I think I successfully installed it for my system settings by manually downloading the 
datrie-0.8.2-cp38-cp38-win_amd64.whl file and renaming it to datrie-0.8.2-cp310-cp310-win_amd64 as pip list shows datrie 0.8.2. However snakemake does not seem to be able to import it and nor can I when I try to import it in my python executable... 

Would love some guidance on this issue. Please let me know what I am doing wrong and if there are any workarounds. 

Thank you so much!

Luc

 - Windows system
 - Python 3.10.14

Paste here what 'pip list' gives you.
Package                   Version     Editable project location
------------------------- ----------- -------------------------
aiobotocore               2.11.2
aiohappyeyeballs          2.4.0
aiohttp                   3.10.5
aioitertools              0.12.0
aiosignal                 1.3.1
anndata                   0.10.9
appdirs                   1.4.4
array_api_compat          1.8
asciitree                 0.3.3
async-timeout             4.0.3
attrs                     24.2.0
botocore                  1.34.19
cellpose                  3.0.11
certifi                   2024.8.30
charset-normalizer        3.3.2
click                     8.1.7
cloudpickle               3.0.0
colorama                  0.4.6
colorcet                  3.1.0
ConfigArgParse            1.7
connection_pool           0.0.3
contourpy                 1.3.0
cycler                    0.12.1
dask                      2024.8.2
dask-expr                 1.1.13
dask-image                2024.5.3
datashader                0.16.3
datrie                    0.8.2
distributed               2024.8.2
docutils                  0.21.2
dpath                     2.2.0
exceptiongroup            1.2.2
fasteners                 0.19
fastjsonschema            2.20.0
fastremap                 1.15.0
fcsparser                 0.2.8
filelock                  3.16.0
flake8                    7.1.1
fonttools                 4.53.1
frozenlist                1.4.1
fsspec                    2023.6.0
geopandas                 1.0.1
gitdb                     4.0.11
GitPython                 3.1.43
h5py                      3.11.0
humanfriendly             10.0
idna                      3.8
imagecodecs               2024.6.1
imageio                   2.35.1
importlib_metadata        8.4.0
Jinja2                    3.1.4
jmespath                  1.0.1
joblib                    1.4.2
jsonschema                4.23.0
jsonschema-specifications 2023.12.1
jupyter_core              5.7.2
kiwisolver                1.4.7
lamin_utils               0.13.4
lazy_loader               0.4
legacy-api-wrap           1.4
llvmlite                  0.43.0
locket                    1.0.0
loompy                    3.0.7
markdown-it-py            3.0.0
MarkupSafe                2.1.5
matplotlib                3.9.2
mccabe                    0.7.0
mdurl                     0.1.2
mpmath                    1.3.0
msgpack                   1.0.8
multidict                 6.0.5
multipledispatch          1.0.0
multiscale_spatial_image  1.0.1
natsort                   8.4.0
nbformat                  5.10.4
networkx                  3.3
numba                     0.60.0
numcodecs                 0.13.0
numpy                     1.26.4
numpy-groupies            0.11.2
ome-zarr                  0.9.0
opencv-python             4.10.0.84
opencv-python-headless    4.10.0.84
packaging                 24.1
pandas                    2.2.2
param                     2.1.1
partd                     1.4.2
patsy                     0.5.6
pillow                    10.4.0
PIMS                      0.7
pip                       24.2
plac                      1.4.3
platformdirs              4.3.2
pooch                     1.8.2
psutil                    6.0.0
PuLP                      2.7.0
pyarrow                   17.0.0
pycodestyle               2.12.1
pyct                      0.5.0
pyflakes                  3.2.0
Pygments                  2.18.0
pynndescent               0.5.13
pyogrio                   0.9.0
pyparsing                 3.1.4
pyproj                    3.6.1
pyreadline3               3.4.1
python-dateutil           2.9.0.post0
pytz                      2024.1
pywin32                   306
PyYAML                    6.0.2
readfcs                   1.1.8
referencing               0.35.1
requests                  2.32.3
reretry                   0.11.8
rich                      13.8.0
roifile                   2024.5.24
rpds-py                   0.20.0
s3fs                      0.6.0
scanpy                    1.10.2
scikit-image              0.24.0
scikit-learn              1.5.1
scipy                     1.14.1
seaborn                   0.13.2
session_info              1.0.0
setuptools                72.1.0
shapely                   2.0.6
shellingham               1.5.4
six                       1.16.0
slicerator                1.1.0
smart-open                7.0.4
smmap                     5.0.1
snakemake                 7.32.4
sopa                      1.1.4       C:\Users\marsdenl\sopa
sortedcontainers          2.4.0
spatial_image             1.1.0
spatialdata               0.2.2
spatialdata-io            0.1.4
statsmodels               0.14.2
stdlib-list               0.10.0
stopit                    1.1.2
sympy                     1.13.2
tabulate                  0.9.0
tblib                     3.0.0
threadpoolctl             3.5.0
throttler                 1.2.2
tifffile                  2024.8.30
toml                      0.10.2
toolz                     0.12.1
toposort                  1.10
torch                     2.4.1
tornado                   6.4.1
tqdm                      4.66.5
traitlets                 5.14.3
typer                     0.12.5
typing_extensions         4.12.2
tzdata                    2024.1
umap-learn                0.5.6
urllib3                   2.0.7
wheel                     0.43.0
wrapt                     1.16.0
xarray                    2024.7.0
xarray-dataclasses        1.8.0
xarray-datatree           0.0.14
xarray-schema             0.0.3
xarray-spatial            0.4.0
yarl                      1.11.0
yte                       1.5.4
zarr                      2.18.3
zict                      3.0.0
zipp                      3.20.1

**Comment by lguerard** [comment]
I also add the issue, and my fix was to do `conda install datrie`. Maybe that helps if you're also using it, otherwise, please discard this message 😄

**Comment by quentinblampey** [comment]
Hello @marsdenl, thanks for reporting this.
Can you let me know if the solution of @lguerard is working?

Else, can you try to install `snakemake<8.1.3`, and in a separate environment? I.e., one environment for sopa, and one environment for Snakemake (then, when running, Snakemake will be able to activate the `sopa` environment).

**Comment by marsdenl** [comment]
Hey @quentinblampey @lguerard that did it, thank you so much! I should have tried this before reaching out, apologies. Managed to download it and load it alright :) 

However, I am having issues with snakemake still (I tried it both on the practice dataset from vizgen and on mine, though they failed at different steps). Here is the code: 

snakemake --config data_path=Z:\Queries\Data\Batch5_region1\region_1 --configfile=Z:\Queries\baysor_vizgen.yaml --cores 1 --use-conda

Here is the log 
Building DAG of jobs...
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job                count
---------------  -------
aggregate              1
all                    1
explorer               1
image_write            1
patchify_baysor        1
report                 1
resolve_baysor         1
to_spatialdata         1
total                  8

Select jobs to execute...

[Tue Sep 10 10:01:36 2024]
rule to_spatialdata:
    input: Z:/Queries/Data/Batch5_region1/region_1
    output: Z:/Queries/Data/Batch5_region1/region_1.zarr/.zgroup
    jobid: 4
    reason: Missing output files: Z:/Queries/Data/Batch5_region1/region_1.zarr/.zgroup
    resources: tmpdir=C:\Users\marsdenl\AppData\Local\Temp, mem_mb=128000, mem_mib=122071

Activating conda environment: sopa
[Tue Sep 10 10:14:36 2024]
Error in rule to_spatialdata:
    jobid: 4
    input: Z:/Queries/Data/Batch5_region1/region_1
    output: Z:/Queries/Data/Batch5_region1/region_1.zarr/.zgroup
    conda-env: sopa
    shell:
        
        sopa read Z:\Queries\Data\Batch5_region1\region_1 --sdata-path Z:\Queries\Data\Batch5_region1\region_1.zarr --technology "merscope"
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Removing output files of failed job to_spatialdata since they might be corrupted:
Z:/Queries/Data/Batch5_region1/region_1.zarr/.zgroup
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: .snakemake\log\2024-09-10T100135.413906.snakemake.log

Would you have any recommendations for this? I assume this is a different issue altogether so happy to close this and create a new one if that's easier for you? 

Thank you very much!

**Comment by quentinblampey** [comment]
Hi, sorry for the delay.
Can you try the command below (i.e., outside of snakemake) and see if you have the same issue? If yes, can you paste the error message?
```sh
sopa read Z:\Queries\Data\Batch5_region1\region_1 --sdata-path Z:\Queries\Data\Batch5_region1\region_1.zarr --technology "merscope"
```

**Comment by marsdenl** [comment]
Sorry I should have closed this issue earlier. I managed to get past this :)

**Comment by quentinblampey** [comment]
Great, thanks! Could you share how you managed to get through this? It could help other users

**Comment by marsdenl** [comment]
To be honest I am not certain why it started working again. When I thought I had initially solved it, I had added single quotation marks around merscope in the .yaml file (technology: 'merscope'). However, I tried to run the code again for this post to make sure it still worked and it did not anymore. So I removed the quotation marks and it worked! The CLI sopa read Z:\Queries\Data\Batch5_region1\region_1 --sdata-path Z:\Queries\Data\Batch5_region1\region_1.zarr --technology "merscope" had always worked, I never had issues with that :) Sorry this isn't very useful for others
